_posts/2015-02-11-abstracting-the-frontend-in-a-shared-nothing-architecture-driven-by-functional-teams.md

```diff
@@ -0,0 +1,167 @@
+      "campaign":"some/fancy/campaign"
+    }
+  }' 
+/>
+{% endhighlight %}
+
+## Standards and conventions ftw
+At this point you might ask: *"Yeah, sounds nice, but what's all the buzz about? This doesn't look like rocket science."* Agreed (mostly). But, that's only a small part of the cake. The majority of work went into defining conventions and standards for the various types and use cases. Which data do we have to pass to our DAL? What are the globally required fields? How do we name our pages? Which data do we need for which event and/or page? What are the more specific bits of information each vertical had to pass? 
+
+The answers to all those questions are very subjective and closely related to the field of business. Obviously, for e-commerce you have substantially different information you want to collect, compared to a content-driven online magazine. But in any case it boils down to some key metrics you want to collect and analyze for your business. E.g. a big part of the DAL's data is passed to our RUM and BI software, which are also implemented as DAL plugins. Thus, many of our own conventions and metrics are specific to this use case. Another major part is the affiliate and basket tracking integration.  
+
+Luckily, I had done most of that long time before, when I initiated the project "tagmanager integration" for our old webshop. So I  started writing down the important KPIs and metrics based on that historical data and the trackingpixels from our tracking solution and our recommendation engine. I summed it all up in a nice table in our wiki, structured the table based on verticals and added a nice description for all keys. Then I wrote the integration stories for the vertical teams so they could integrate the right metatags and declarations in the markup. Awesome.
+
+Well, at least it felt awesome until I looked at the actual implementation done by the teams. The problem was that I haven't been very explicit in declaring the types (and especially the formatting) for the individual metrics. I simply defined something like *on the product detail page we need a product object with the following fields: name, price, category, ...*. Even though I also defined some examples, this still led to fairly diverse implementations. Especially the price formatting was a problem, because there were multiple interpretations about how a price should be expressed (e.g.`"29,90€"` vs. `29.9` which are both valid representations of the same price value).
+
+## Typization to the rescue
+Since Javascript (and JSON, too) is a very loosely typed language, we needed to define abstract types (or interfaces if you're using [Typescript](http://typescriptlang.org)) that explicitly define how values have to be supplied to the DAL. That resulted in various fancy types, e.g. *DALPageData*, *DALUserData* or *DALProductData* to just name a few. Let's look at the type definition for the *DALPageData* type:
+
+{% highlight javascript %}
```

Comment:
Could make sense to give an example for the price representation, because you talked about it in the previous paragraph.